tests/Core/Content/Product/SalesChannel/Listing/ProductListingFeaturesSubscriberTest.php

```diff
@@ -893,4 +895,73 @@ private function getDefaultSessionMock(): SessionInterface
 
         return $sessionMock;
     }
+
+    public function criteriaLimitProvider(): array
+    {
+        return [
+            'search request' => [
+                'endpoint' => 'handleSearchRequest',
+                'expectedLimit' => 3,
+                'isNavigationRequest' => false
+            ],
+            'navigation request' => [
+                'endpoint' => 'handleListingRequest',
+                'expectedLimit' => 3,
+                'isNavigationRequest' => true
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider criteriaLimitProvider
+     */
+    public function testCriteriaLimitIsSetForPagination(
+        string $endpoint,
+        int $expectedLimit,
+        bool $isNavigationRequest
+    ): void {
+        if (!$isNavigationRequest) {
+            $eventMock = $this->setUpSearchRequestMocks($this->getDefaultResponse());
+        } else {
+            $eventMock = $this->setUpNavigationRequestMocks();
+        }
+
+        $expectedAssign = [
+            'sorting' => [],
+            'filters' => [],
+            'postFilters' => [],
+            'aggregations' => [],
+            'queries' => [],
+            'groupFields' => [],
+            'offset' => null,
+            'limit' => null,
+            'totalCountMode' => 0,
+            'associations' => [],
+            'ids' => [
+                '019111105-37900' => '019111105-37900',
+                '029214085-37860' => '029214085-37860'
+            ],
+            'states' => [],
+            'inherited' => false,
+            'term' => null,
+            'extensions' => [
+                'flPagination' => new Pagination($expectedLimit, 0, 1808)
+            ],
+            'includes' => null
+        ];
+        if (Utils::versionLowerThan('6.3')) {
+            $expectedAssign['source'] = null;
+        }
+        $expectedAssign['title'] = null;
+
+        $criteriaMock = $this->getMockBuilder(Criteria::class)->disableOriginalConstructor()->getMock();
+        $criteriaMock->expects($this->any())->method('assign')->with($expectedAssign);
```

Comment:
If you use `$this->any()` in combination with a `with()`, you are never sure that this has ever been called, and you can not be sure that the data has been properly set. I suggest using a real criteria and asserting the actual data of it after calling the subscriber search endpoint.

In fact I am sure that this test must fail, since the required change has not been added to the `CompatibilityLayer` classes, and yet the test still passes, which should not be the case.